### DISCUSSION TITLE
Overwrite stdio functions

### DISCUSSION THREAD

**Discussion by akaeba** [charlesnicholson/nanoprintf#231] [Ideas]
Hi Charles,

the nanoprintf.h is a real great deal. Thanks for providing such an powerful tool. I have on addition in mind, to replace the stdio printf implementation by the npf_printf implementation. Enabling could be realized by compile switch. What do you think? I started a short google research: 
* https://en.cppreference.com/w/cpp/preprocessor/replace
* http://gcc.gnu.org/onlinedocs/gcc/Variadic-Macros.html

Without testing it could look like follows:
```c
#undef printf
#define printf(format, ...) npf_pprintf (format, ## __VA_ARGS__)
```

BR,
Andreas

**Comment by charlesnicholson** [comment]
Thanks for the kind words!

The C99 Standard defines printf (and friends) explicitly as a _function_- unfortunately `#undef` doesn't work on functions, only preprocessor symbols. 

<img width="664" alt="Screenshot 2022-11-19 at 3 15 58 PM" src="https://user-images.githubusercontent.com/3010295/202869848-a97febfd-7766-42e0-879b-66627faea6f8.png">

I guess there could be a compile-time flag into `nanoprintf` that renames functions, but I think there's still a problem:

Under the hood `printf` writes the content to the `stdout` io stream. `printf(char const *fmt, ...)` is generally implemented as `fprintf(stdout, fmt, ...)` for this reason; it's a shortcut. I could perhaps see it being helpful to have a compile-time build flag for `nanoprintf` that turns `npf_snprintf` into `snprintf`, and similarly turns `npf_vsnprintf` into `vsnprintf`. Though, if you simply write your own `snprintf` and `vsnprintf` wrappers that just return whatever the `npf_` versions return, the compiler will optimize the shims out of existence.

Since `nanoprintf` is intended to replace stdio, though, I'd prefer that it not depend on stdio so that it can have access to the `stdout` symbol. Additionally, the per-byte output callback would have to write to the `stdout` stream, which means that `nanoprintf` would take a dependency on `write`.

My intent with `nanoprintf` was to be a building block for users who wanted to replace or re-implement `printf`, and to that end I think a big part of its power and flexibility come from its willful ignorance of `write` and `stdout`. To me it's one of those "less is more" scenarios :)

Anyway, I hope that was helpful!

**Reply by akaeba** [reply]
Hi Charles,
thanks for the explanaition. It looks much complexer then i thought at begin. Are all valid arguments. I'll try it in my embedded project to overwrite the function. Thanks :)

BR,
Andreas
